### PR TITLE
[IMM32] Use Imm32IsImmMode and Imm32IsCiceroMode macros

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Define `Imm32IsImmMode`, `Imm32IsCiceroMode` and `Imm32Is16BitMode` macros in `"precomp.h"`.
- Use them.
- Improve `Imm32InquireIme` and `Imm32CleanupContext` functions.
- Add `IME_SYSINFO_WINLOGON` and `IME_SYSINFO_WOW16` macros (from DDK) in `psdk/imm.h`.
- Modify `CtfImeInquireExW` and `CtfImeSelectEx` prototype.

### CODE REVIEW

File: dll/win32/imm32/precomp.h [hunk 1]
```diff
@@ -93,7 +93,14 @@ static inline BOOL Imm32IsCrossProcessAccess(HWND hWnd)
             (DWORD_PTR)NtCurrentTeb()->ClientId.UniqueProcess);
 }
 
-#define ImeDpi_IsUnicode(pImeDpi) ((pImeDpi)->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+#define ImeDpi_IsUnicode(pImeDpi)   ((pImeDpi)->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+#define Imm32IsImmMode()            (g_psi && (g_psi->dwSRVIFlags & SRVINFO_IMM32))
+#define Imm32IsCiceroMode()         (g_psi && (g_psi->dwSRVIFlags & SRVINFO_CICERO_ENABLED))
+
+static inline BOOL Imm32IsSomethingMode(VOID) /* TODO: unknown */
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:100]: why not using a macro too?

File: dll/win32/imm32/precomp.h [hunk 2]
```diff
@@ -93,7 +93,14 @@ static inline BOOL Imm32IsCrossProcessAccess(HWND hWnd)
...
+{
+    return !!(((PW32CLIENTINFO)NtCurrentTeb()->Win32ClientInfo)->W32ClientInfo[0] & 2);
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:102]: This is the check of `TIF_16BIT` in the thread's `THREADINFO::TIF_flags`
See https://reactos.org/wiki/Techwiki:Win32k/THREADINFO and https://www.geoffchappell.com/studies/windows/km/win32k/structs/threadinfo/tif_flags.htm
and our user32.

File: dll/win32/imm32/precomp.h (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: OK.

File: dll/win32/imm32/precomp.h (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: I had to move some structures from `win32.h` to `ntuser.h`...

Comment (HBelusca) [reply dll/win32/imm32/precomp.h]: Why? user32 code can access them as they currently are 🤔

Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: 1. We cannot include `win32.h`.

File: win32ss/user/ntuser/win32.h [hunk 3]
```diff
@@ -52,120 +50,6 @@ extern HANDLE hModuleWin;    // This Win32k Instance.
 extern struct _CLS *SystemClassList;
 extern BOOL RegisteredSysClasses;
 
-#include <pshpack1.h>
-// FIXME: Move to ntuser.h
```
Comment (katahiromz) [inline win32ss/user/ntuser/win32.h:56]: 2. There were FIXME that argues to move the structures.

Comment (katahiromz) [reply win32ss/user/ntuser/win32.h:56]: 3. When I moved these, compilation passed.

File: dll/win32/imm32/precomp.h [hunk 4]
```diff
@@ -93,7 +93,11 @@ static inline BOOL Imm32IsCrossProcessAccess(HWND hWnd)
             (DWORD_PTR)NtCurrentTeb()->ClientId.UniqueProcess);
 }
 
-#define ImeDpi_IsUnicode(pImeDpi) ((pImeDpi)->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+#define ImeDpi_IsUnicode(pImeDpi)   ((pImeDpi)->ImeInfo.fdwProperty & IME_PROP_UNICODE)
+#define Imm32IsImmMode()            (g_psi && (g_psi->dwSRVIFlags & SRVINFO_IMM32))
+#define Imm32IsCiceroMode()         (g_psi && (g_psi->dwSRVIFlags & SRVINFO_CICERO_ENABLED))
+#define Imm32Is16BitMode() \
+    (((PTHREADINFO)NtCurrentTeb()->Win32ThreadInfo)->TIF_flags & TIF_16BIT)
```
Comment (HBelusca) [inline dll/win32/imm32/precomp.h:100]:
```suggestion
    (GetWin32ClientInfo()->dwTIFlags & TIF_16BIT)
```
The `THREADINFO` structure contents are inaccessible from usermode.

File: win32ss/user/ntuser/win32.h (hunk 3 above)
Comment (HBelusca) [reply win32ss/user/ntuser/win32.h:56]: This comment was erroneous. These structures are only valid in kernel mode (within win32k).

File: win32ss/include/ntuser.h [hunk 5]
```diff
@@ -3696,6 +3698,119 @@ NtUserSetScrollBarInfo(
     LONG idObject,
     SETSCROLLBARINFO *info);
 
+#include <pshpack1.h>
```
Comment (HBelusca) [inline win32ss/include/ntuser.h:3701]: Please revert (see comment below)

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: win32ss/user/ntuser/win32.h (hunk 3 above)
Comment (katahiromz) [reply win32ss/user/ntuser/win32.h]: OK.

File: win32ss/include/ntuser.h (hunk 5 above)
Comment (katahiromz) [reply win32ss/include/ntuser.h]: Reverted `ntuser.h`.

File: dll/win32/imm32/precomp.h (hunk 4 above)
Comment (katahiromz) [reply dll/win32/imm32/precomp.h]: I see.

**Review by HBelusca — APPROVED** [review]

**Pull request merged**